File: src/main/java/com/mbayou/kick4k/chat/PostChatMessageRequest.java
package com.mbayou.kick4k.chat;

import com.fasterxml.jackson.annotation.JsonInclude;
import com.fasterxml.jackson.annotation.JsonValue;

@JsonInclude(JsonInclude.Include.NON_NULL)
public class PostChatMessageRequest {
    private final Integer broadcasterUserId;
    private final String content;
    private final String replyToMessageId;
    private final Type type;

    public PostChatMessageRequest(Integer broadcasterUserId, String content, String replyToMessageId, Type type) {
        this.broadcasterUserId = broadcasterUserId;
        this.content = content;
        this.replyToMessageId = replyToMessageId;
        this.type = type;
    }

    public Integer getBroadcasterUserId() {
        return broadcasterUserId;
    }

    public String getContent() {
        return content;
    }

    public String getReplyToMessageId() {
        return replyToMessageId;
    }

    public Type getType() {
        return type;
    }

    public static Builder builder() {
        return new Builder();
    }

    public static class Builder {
        private Integer broadcasterUserId;
        private String content;
        private String replyToMessageId;
        private Type type;

        public Builder broadcastUserId(Integer broadcastUserId) {
            this.broadcasterUserId = broadcastUserId;
            return this;
        }

        public Builder content(String content) {
            this.content = content;
            return this;
        }

        public Builder replyToMessageId(String replyToMessageId) {
            this.replyToMessageId = replyToMessageId;
            return this;
        }

        public Builder type(Type type) {
            this.type = type;
            return this;
        }

        public PostChatMessageRequest build() {
            if (content == null)  {
                throw new IllegalStateException("Content is required");
            }

            if (type == null) {
                throw new IllegalStateException("Type is required");
            }

            if (type == Type.USER && broadcasterUserId == null) {
                throw new IllegalStateException("BroadcastUserId is required when sending as user");
            }

            return new PostChatMessageRequest(broadcasterUserId, content, replyToMessageId, type);
        }
    }

    public enum Type {
        USER, BOT;

        @JsonValue
        public String toLower() {
            return name().toLowerCase();
        }
    }
}

File: src/main/kotlin/com/mbayou/kick4k/chat/PostChatMessageRequest.kt
package com.mbayou.kick4k.chat

import com.fasterxml.jackson.annotation.JsonInclude
import com.fasterxml.jackson.annotation.JsonValue

@JsonInclude(JsonInclude.Include.NON_NULL)
data class PostChatMessageRequest(
    val broadcasterUserId: Int?,
    val content: String,
    val replyToMessageId: String?,
    val type: Type,
) {
    companion object {
        @JvmStatic
        fun builder(): Builder = Builder()
    }

    class Builder {
        private var broadcasterUserId: Int? = null
        private var content: String? = null
        private var replyToMessageId: String? = null
        private var type: Type? = null

        fun broadcastUserId(broadcastUserId: Int?) = apply { this.broadcasterUserId = broadcastUserId }
        fun content(content: String) = apply { this.content = content }
        fun replyToMessageId(replyToMessageId: String?) = apply { this.replyToMessageId = replyToMessageId }
        fun type(type: Type) = apply { this.type = type }

        fun build(): PostChatMessageRequest {
            val contentValue = requireNotNull(content) { "Content is required" }
            val typeValue = requireNotNull(type) { "Type is required" }
            if (typeValue == Type.USER && broadcasterUserId == null) {
                throw IllegalStateException("BroadcastUserId is required when sending as user")
            }
            return PostChatMessageRequest(broadcasterUserId, contentValue, replyToMessageId, typeValue)
        }
    }

    enum class Type {
        USER,
        BOT;

        @JsonValue
        fun toLower(): String = name.lowercase()
    }
}
